keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -40,74 +56,145 @@ For operators who want to regularly rotate the signing and verifying keys for pr
+- cumulation on local and external keys will not be supported.
+- If Both, `--service-account-key-file` and `--key-service-url` are specified, keys from external signers will override the keys at `service-account-key-file` as supported public keys.
+
+### Risks and Mitigations
 
-package v1alpha1;
+- New token generation and validation could suffer a performance difference when reading over a socket, as an external process will be signing data.
+- Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket.
+- External signers will only return signature and header while the token will be assembled by TokenGenerator implementation for external signing. This will help ApiServer control the payload and thus the claims.
 
+## Design Details
+
+### KeyService RPC
+
+will be served on a socket as configured via the `--key-service-url` flag.
+
+```proto
+// This service is served by a process on a local Unix Domain Socket.
 service KeyService {
-  // Sign an incoming payload
-  rpc SignPayload(SignPayloadRequest) returns (SignPayloadResponse) {}
-  // List all active public keys
-  rpc ListPublicKeys(ListPublicKeysRequest) returns (ListPublicKeysResponse) {}
+  // Sign takes a serialized JWT payload, and returns the serialized header and
+  // signature.  The caller can then assemble the JWT from the header, payload,
+  // and signature.
+  //
+  // The plugin MUST set a key id in the returned JWT header.
+  rpc Sign(SignJWTRequest) returns (SignJWTResponse) {}
+
+  // FetchKeys returns the set of public keys that are trusted to sign
+  // Kubernetes service account tokens. Kube-apiserver will call this RPC:
+  //
+  // * Every time it tries to validate a JWT carrying an unknown key ID, and
+  //
+  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
+  //   JWKs endpoint.
+  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}
 }
 
-message SignPayloadRequest {
-  // payload is the content to be signed. JWT headers must be included by the caller
-  bytes payload = 1;
-  // algorithm specifies which algorithm to sign with
-  string algorithm = 2;
+message SignJWTRequest {
+  // The raw payload to be signed.  A serialized JSON object, no base64 wrapping.
+  optional bytes payload = 1;
 }
-message SignPayloadResponse {
-  // content returns the signed payload
-  bytes content = 1;
+
+message SignJWTResponse {
+  // The header for the JWT.  Already wrapped in URL-safe base64, valid for
+  // directly inserting into the final JWT.
+  optional bytes header = 1;
+
+  // The signature for the JWT.  Already wrapped in URL-safe base64, valid for
+  // directly inserting into the final JWT.
+  optional bytes signature = 2;
 }
 
+message FetchKeysRequest {
+    optional string triggering_key_id = 1;
```

Comment:
Consensus in the discussion was to drop this field until there was a concrete use case.

Reply:
done